src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/Algebras/BDDAlgebra.cs

```diff
@@ -90,58 +92,69 @@ public BDD GetOrCreateBDD(int ordinal, BDD? one, BDD? zero)
-                return a;
-
-            if (a == True || b == True)
-                return True;
-
-            if (a == b)
-                return a;
-
-            BoolOpKey key = CreateBoolOpKey(BoolOp.Or, a, b);
-            return _binOpCache.TryGetValue(key, out BDD? res) ?
-                res :
-                CreateBoolOP_lock(key);
-        }
+        public BDD Or(BDD a, BDD b) =>
+            a == False ? b :
+            b == False ? a :
+            a == True || b == True ? True :
+            a == b ? a :
+            GetOrCreateBoolOp(CreateBinOpKey(BoolOp.Or, a, b));
 
         /// <summary>
         /// Make the intersection of a and b
         /// </summary>
-        public BDD And(BDD a, BDD b)
-        {
-            if (a == True)
-                return b;
-
-            if (b == True)
-                return a;
-
-            if (a == False || b == False)
-                return False;
-
-            if (a == b)
-                return a;
-
-            BoolOpKey key = CreateBoolOpKey(BoolOp.And, a, b);
-            return _binOpCache.TryGetValue(key, out BDD? res) ?
-                res :
-                CreateBoolOP_lock(key);
-        }
+        public BDD And(BDD a, BDD b) =>
+            a == True ? b :
+            b == True ? a :
+            a == False || b == False ? False :
+            a == b ? a :
+            GetOrCreateBoolOp(CreateBinOpKey(BoolOp.And, a, b));
 
         /// <summary>
         /// Complement a
         /// </summary>
         public BDD Not(BDD a) =>
             a == False ? True :
             a == True ? False :
-            _notCache.TryGetValue(a, out BDD? neg) ? neg :
-            CreateBoolOP_lock(new BoolOpKey(BoolOp.Not, a, null));
+            GetOrCreateBoolOp(new BoolOpKey(BoolOp.Not, a, null));
+
+        private BDD GetOrCreateBoolOp(BoolOpKey key)
+        {
+            Lock.EnterReadLock();
+            try
+            {
+                if (_opCache.TryGetValue(key, out BDD? result))
+                {
+                    return result;
+                }
+            }
+            finally
+            {
+                Lock.ExitReadLock();
+            }
+            // No result was found in cache
+            Lock.EnterUpgradeableReadLock();
+            try
+            {
+                // Check again for entry in cache, since another thread may have created it when read lock was released
+                if (!_opCache.TryGetValue(key, out BDD? result))
+                {
+                    Lock.EnterWriteLock();
+                    try
+                    {
+                        result = CreateBoolOP_lock(key);
+                    }
+                    finally
+                    {
+                        Lock.ExitWriteLock();
+                    }
+                }
+                return result;
+            }
+            finally
+            {
+                Lock.ExitUpgradeableReadLock();
+            }
+        }
```

Comment:
Similar question here.  I'm wondering if it might be better for _opCache to be a ConcurrentDictionary; then we wouldn't lock while doing reads, and would just use a normal local if we had a cache miss, e.g.
```C#
if (!_opCache.TryGetValue(key, out BDD? result)
{
    lock (_opCache)
    {
        if (!_opCache.TryGetValue(key, out result)
        {
            result = CreateBoolOp_lock(key);
        }
    }
}
return result;
```

Reply:
I talked with Margus about this too and I think he was convinced it should work, I'll implement the fine grained synchronization.

Reply:
>  with all updates to the cache using GetOrAdd

If from a correctness perspective it's ok to have the same key being processed concurrently,  I agree this is the way to go.